### ISSUE TITLE
bug(android/common): predictive text - fatal error when loading the EuroLatin keyboard

### ISSUE THREAD

**Issue by MakaraSok** [keymanapp/keyman#4647]
**Describe the bug**

When loading EuroLatin keyboard on Android 5.0, a fatal error occurs.

After a fresh install completed:
![Screenshot_1615433722](https://user-images.githubusercontent.com/28331388/110731809-a1019a80-8255-11eb-85e1-95c443f81a14.png)

From Keyman web console:
![image](https://user-images.githubusercontent.com/28331388/110730294-cb9e2400-8252-11eb-8e94-aa32dc458523.png)

**To Reproduce**

1. Fresh install keyman 14.0.257-beta
2. See error
3. Close the 'Get Started'
4. Install a new keyboard - anyone
5. Switch between the two keyboards 
6. See error when switch back to EuroLatin

**Expected behavior**

The predictive text should be loaded and work as expected without the fatal error.

---

**Keyman for Android:**
 - Device: Emulator Pixel_4_API_21
 - OS: Android 5.0.2 x86_64
 - Keyman version 14.0.257-beta
 - Target application: Keyman
 - Android system webview: 37.0

**KeymanWeb:**
 - OS: Windows 10 Pro 20H2 x64
 - Browser: Chrome Version 89.0.4389.82 (Official Build) (64-bit)
 - KeymanWeb Version: ?
 - URL: N/A

---

**Keyboard**
 - Keyboard name: EuroLatin (SIL)
 - Keyboard version: 1.9.3
 - Language name: English

**Comment by darcywong00** [comment]
Hm, it loads fine in my Android Emulator with:
Lollipop API 21, ABI x86 Target Android 5.0

**Comment by jahorton** [comment]
According to the Sentry issue, it's not _new_; it's something that appears to have been happening rarely for a while.  Though yeah, it's fairly specific to Android 5.0, from what Sentry can see.

![Screen Shot 2021-03-11 at 11 33 42 AM](https://user-images.githubusercontent.com/25213402/110736115-aa8f0080-825d-11eb-8d74-5ecf4761f621.png)

**Comment by darcywong00** [comment]
Most might be coming from @MakaraSok 's emulator, so I'll have him try to rebuild the Android app and see if we can get a better stack trace

**Comment by darcywong00** [comment]
Are there arrow functions for loading the dictionary?

If so, it needs Chrome 45
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions
while Android 5.0 has Chrome 37

**Comment by jahorton** [comment]
> Are there arrow functions for loading the dictionary?
> 
> If so, it needs Chrome 45
> https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions
> while Android 5.0 has Chrome 37

While that shouldn't result in the exact error you're seeing, that might actually be relevant.

From MTNT 0.1.6:

```
      searchTermToKey: function defaultCasedSearchTermToKey(wordform, applyCasing) {
      // While this is a bit WET, as the basic `defaultSearchTermToKey` exists and performs some of
      // the same functions, repetition is the easiest way to allow the function to be safely compiled 
      // with ease by use of `.toString()`.
      return Array.from(wordform
          .normalize('NFKD')
          // Remove any combining diacritics (if input is in NFKD)
          .replace(/[\u0300-\u036F]/g, '')) // end of `Array.from`
          .map(c => applyCasing('lower', c))
          .join('');
  }
```

There _is_ an arrow-function in the bundled `searchTermToKey` function.  If you crack open that KMP and replace the offending line with:

```
          .map(function(c) { return applyCasing('lower', c) })
```

That small change should remove the influence of arrow functions and yield some diagnostic insight.

**Comment by darcywong00** [comment]
I've made that change and sent to @MakaraSok for him to test in his emulator. 

I did notice nearly all the lexical-models use that map for applyCasing...

**Comment by jahorton** [comment]
After some more investigation and analysis, my best guess is now that #4370's changes may be responsible.

There's a brief period during model loading where it's set as active _before_ the model is loaded.  If the load fails, it's then unloaded.  While I don't see where this is directly calling the `unload` method... at least not yet... there may be something that's triggering it after a failed load attempt.

https://github.com/keymanapp/keyman/blob/ae6068557eefed62e9fb87ea07787744ec583e70/common/core/web/input-processor/src/text/prediction/languageProcessor.ts#L140-L163

I'll continue investigating this for a bit.

**Comment by jahorton** [comment]
There's a chance this block may be important:

https://github.com/keymanapp/keyman/blob/ae6068557eefed62e9fb87ea07787744ec583e70/web/source/text/prediction/modelManager.ts#L35-L43

There's no promise-waiting between the `unload` and the `load` calls, so that may be a minor weak point?

**Comment by jahorton** [comment]
So it's more visible, the full-detailed Sentry report did have a bit more info that helped lead me in this direction:

![Screen Shot 2021-03-17 at 3 00 23 PM](https://user-images.githubusercontent.com/25213402/111434001-9782a180-8731-11eb-95c5-2d20cf60aeb8.png)

**Comment by mcdurdin** [comment]
This bug will be fixed by #4716.

**Comment by mcdurdin** [comment]
I think the expected load but got unload happens when you reload the page with a model that failed to load.

**Comment by darcywong00** [comment]
Fixed by PR #4718

**Comment by MakaraSok** [comment]
Retested with 14.0.266-beta. Confirm that the error has been fixed. The predictive text for the default keyboard (EuroLatin (SIL)) is working without any issue.
